presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java

```diff
@@ -56,17 +80,118 @@ public MemoryRevokingScheduler(LocalMemoryManager localMemoryManager, FeaturesCo
         this.memoryPools = ImmutableList.of(memoryPool);
         this.memoryRevokingThreshold = memoryRevokingThreshold;
         this.memoryRevokingTarget = memoryRevokingTarget;
+        this.potentiallyOverflowingPools = synchronizedSet(Sets.newHashSetWithExpectedSize(this.memoryPools.size()));
+    }
+
+    @PostConstruct
+    public void start()
+    {
+        memoryPools.forEach(memoryPool -> memoryPool.addListener(memoryPoolListener));
+    }
+
+    @PreDestroy
+    public void stop()
+    {
+        memoryPools.forEach(memoryPool -> memoryPool.removeListener(memoryPoolListener));
+    }
+
+    void register(Supplier<? extends Collection<SqlTask>> currentTasks, ScheduledExecutorService taskManagementExecutor)
+    {
+        checkState(this.currentTasks == null, "Already registered");
+        this.currentTasks = requireNonNull(currentTasks, "currentTasks cannot be null");
+        this.taskManagementExecutor = requireNonNull(taskManagementExecutor, "taskManagementExecutor cannot be null");
+
+        taskManagementExecutor.scheduleWithFixedDelay(() -> {
+            try {
+                checkMemoryPools();
+            }
+            catch (Throwable e) {
+                log.warn(e, "Error running period memory revocation check");
+            }
+        }, 0, 1, TimeUnit.SECONDS);
+    }
+
+    private void onMemoryReserved(MemoryPool memoryPool)
+    {
+        try {
+            if (!memoryRevokingNeeded(memoryPool, memoryPool.getFreeBytes())) {
+                return;
+            }
+
+            if (potentiallyOverflowingPools.add(memoryPool.getId())) {
+                taskManagementExecutor.execute(() -> {
+                    try {
+                        checkMemoryPools(false);
+                    }
+                    catch (Throwable e) {
+                        log.warn(e, "Error requesting immediate memory revoking");
+                    }
+                });
+            }
+        }
+        catch (Throwable e) {
+            log.warn(e, "Error when acting on memory pool reservation");
+        }
     }
 
-    public void requestMemoryRevokingIfNeeded(Collection<SqlTask> sqlTasks)
+    @VisibleForTesting
+    void checkMemoryPools()
     {
-        memoryPools.forEach(memoryPool -> requestMemoryRevokingIfNeeded(sqlTasks, memoryPool));
+        checkMemoryPools(true);
+    }
+
+    private void checkMemoryPools(boolean checkAllPools)
+    {
+        /*
+         * There can be at most one thread actually executing this method. If pool or pools need to be re-checked,
+         * this will be signalled by setting immediateCheckPending and populating potentiallyOverflowingPools.
+         */
+        if (!checkRunning.compareAndSet(false, true)) {
+            // Already running
+
+            if (checkAllPools) {
```

Comment:
The reason is to avoid adding/removing the pools to the Set every 1 s when nothing else happens. I.e. to keep low profile when idle.